### PR TITLE
For ipvlan tests check that the ipvlan module is enabled

### PR DESCRIPTION
**Pull request by jim-docker**
**- What I did**
Made sure tests that require ipvlan module to be enabled are run only if it is enabled

**- How I did it**
Added Kir's test for ipvlan module being enabled

**- How to verify it**
Run TestDockerNetworkIpvlan test with and without ipvlan being enabled

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

File: integration/network/ipvlan/ipvlan_test.go [hunk 1]
```diff
@@ -425,7 +429,26 @@ func testIpvlanAddressing(client dclient.APIClient) func(*testing.T) {
...
-// ensure Kernel version is >= v4.2 for ipvlan support
+var (
+	once            sync.Once
+	ipvlanSupported bool
+)
+
+// figure out if ipvlan is supported by the kernel
 func ipvlanKernelSupport() bool {
-	return n.CheckKernelMajorVersionGreaterOrEqualThen(4, 2)
+	if n.CheckKernelMajorVersionGreaterOrEqualThen(4, 2) == false {
```
Comment (thaJeztah) [inline integration/network/ipvlan/ipvlan_test.go]: Generally we don't compare with a boolean, so probably;

```suggestion
	if !n.CheckKernelMajorVersionGreaterOrEqualThen(4, 2) {
```

Perhaps we should add a `CheckKernelMajorVersionLowerThan`

File: integration/network/ipvlan/ipvlan_test.go [hunk 2]
```diff
@@ -425,7 +429,26 @@ func testIpvlanAddressing(client dclient.APIClient) func(*testing.T) {
...
+		return false
+	}
+	once.Do(func() {
+		// this may have the side effect of enabling the ipvlan module
+		exec.Command("modprobe", "ipvlan").Run()
+		_, err := os.Stat("/sys/module/ipvlan")
+		if err == nil {
+			ipvlanSupported = true
+		} else if !os.IsNotExist(err) {
+			fmt.Printf("WARNING: ipvlanKernelSupport: stat failed: %v\n", err)
```
Comment (thaJeztah) [inline integration/network/ipvlan/ipvlan_test.go]: should we pass `testing.T` to this function, and `t.Logf()` here?

**Review by thaJeztah — CHANGES_REQUESTED** [review]

**Comment by thaJeztah** [comment]
ping @kolyshkin ptal

File: integration/network/ipvlan/ipvlan_test.go (hunk 1 above)
Comment (thaJeztah) [reply integration/network/ipvlan/ipvlan_test.go]: Ah, looks like the linter found it as well 

```
21:11:48 integration/network/ipvlan/ipvlan_test.go:439:5:warning: should omit comparison to bool constant, can be simplified to !n.CheckKernelMajorVersionGreaterOrEqualThen(4, 2) (S1002) (gosimple)

File: integration/network/ipvlan/ipvlan_test.go (hunk 2 above)
Comment (jim-docker) [reply integration/network/ipvlan/ipvlan_test.go]: Yes, or return an error?

Comment (thaJeztah) [reply integration/network/ipvlan/ipvlan_test.go]: It's used as a boolean to skip a test, so I'm not sure that'd help.

If we want to _fail_ in this situation, we could do `t.Errorf()`

Possibly this function should be marked as a `t.Helper()`, but I'm ok with leaving that for now.

Comment (jim-docker) [reply integration/network/ipvlan/ipvlan_test.go]: I meant return an error as well as the boolean. This function is fairly generic (albeit it is not in a generic package) so it seems a little forced to pass it `testing.T`

Comment (thaJeztah) [reply integration/network/ipvlan/ipvlan_test.go]: I'm fine either way; it's just that it's implemented as a helper for tests (so messages should ideally be printed through the testing code), and returning an error would mean we'd have to use an intermediate `bool` in the test itself, and handle potential errors there

**Comment by kolyshkin** [comment]
Two things

1. No need to check for the kernel version at all, as we already have a better check. Some distros might have backported this functionality to older kernel (and in general these days kernel version doesn't mean much so we only check it as a last resort, i.e. if we have no other way).

2. Please squash your commits, it's very hard to review it the way it is.

**Comment by jim-docker** [comment]
> Two things
> 
> 1. No need to check for the kernel version at all, as we already have a better check. Some distros might have backported this functionality to older kernel (and in general these days kernel version doesn't mean much so we only check it as a last resort, i.e. if we have no other way).
> 2. Please squash your commits, it's very hard to review it the way it is.

1) OK. I had the impression the problem was with the kernel < 4.2, even if ipvlan is enabled. If not I will remove the version check.

2) squashing now will make a mess of other reviews, no? Don't you just review the diff between my branch and moby master?

**Comment by kolyshkin** [comment]
> I had the impression the problem was with the kernel < 4.2, even if ipvlan is enabled. If not I will remove the version check.

Does not matter. We now have better check for the feature, and it's sufficient, so there's absolutely no reason to wrap it into another, much less reliable, check.

> Don't you just review the diff

I can and I actually did, as per the comment about kernel version check above.

We still need to squash it anyway before merging, as it's one small logical change and it should be one commit -- otherwise we'll pollute git log

**Comment by jim-docker** [comment]
> We still need to squash it anyway before merging, as it's one small logical change and it should be one commit -- otherwise we'll pollute git log
I'll squash it, but I recall seeing a checkbox you can select when merging to master to do the squash automatically

**Comment by thaJeztah** [comment]
> I'll squash it, but I recall seeing a checkbox you can select when merging to master to do the squash automatically

We tend not to use that option because;

- it modifies the contributor's commit
- commits that were GPG signed by the contributor no longer are signed
- commits can no longer traced back (or at least; no longer match) the commits in the pull request

**Review by thaJeztah — APPROVED** [review]
LGTM

**Review by kolyshkin — APPROVED** [review]
LGTM

**Pull request merged**